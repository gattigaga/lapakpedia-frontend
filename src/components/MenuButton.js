import React, { Component, createRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TweenLite, Power2 } from "gsap";

const Container = styled.svg`
  width: 32px;
  height: 32px;
  cursor: pointer;
`;

const Line = styled.line`
  stroke-width: 4px;
  stroke: black;
`;

class MenuButton extends Component {
  constructor(props) {
    super(props);

    this.lines = [...Array(3)].map(createRef);
  }

  componentDidMount() {
    const { isOpen } = this.props;

    this.handleAnimation(isOpen);
  }

  componentWillReceiveProps(nextProps) {
    const { isOpen } = this.props;

    if (isOpen !== nextProps.isOpen) {
      this.handleAnimation(nextProps.isOpen);
    }
  }

  /**
   * Handle open and close state animation
   *
   * @param {boolean} isOpen - Check if user wants menu button to open or not
   * @memberof MenuButton
   */
  handleAnimation(isOpen) {
    const getY = index => 32 / 4 * (index + 1);

    TweenLite.to(this.lines[0].current, 0.3, {
      ease: Power2.easeInOut,
      attr: {
        y1: getY(isOpen ? 1 : 0),
        y2: getY(isOpen ? 1 : 0)
      }
    });

    TweenLite.to(this.lines[2].current, 0.3, {
      ease: Power2.easeInOut,
      attr: {
        y1: getY(isOpen ? 1 : 2),
        y2: getY(isOpen ? 1 : 2)
      }
    });
  }

  render() {
    const { onClick } = this.props;

    return (
      <Container onClick={onClick}>
        {[...Array(3)].map((line, index) => {
          const y = `${32 / 4 * (index + 1)}px`;

          return (
            <Line
              key={index}
              x1="4px"
              y1={y}
              x2="28px"
              y2={y}
              innerRef={this.lines[index]}
            />
          );
        })}
      </Container>
    );
  }
}

MenuButton.propTypes = {
  isOpen: PropTypes.bool,
  onClick: PropTypes.func
};

MenuButton.defaultProps = {
  isOpen: false
};

export default MenuButton;
