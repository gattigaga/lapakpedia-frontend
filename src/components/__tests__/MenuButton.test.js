import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";

import MenuButton from "components/MenuButton";

describe("MenuButton", () => {
  const setup = propOverrides => {
    const props = {
      ...propOverrides
    };

    const wrapper = mount(<MenuButton {...props} />);

    return {
      props,
      wrapper
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should in open state", () => {
    const { wrapper } = setup({ isOpen: true });

    expect(wrapper.props().isOpen).toEqual(true);
  });

  it("should call 'onClick' callback while clicked", () => {
    const onClick = jest.fn();
    const { wrapper } = setup({ onClick });

    wrapper.simulate("click");

    expect(onClick).toBeCalled();
  });
});
