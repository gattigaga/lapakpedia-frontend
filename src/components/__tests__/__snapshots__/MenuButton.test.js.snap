// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuButton should renders default 1`] = `
<MenuButton
  isOpen={false}
>
  <styled.svg>
    <svg
      className="sc-bdVaJa dPqvbs"
    >
      <styled.line
        innerRef={
          Object {
            "current": <line
              class="sc-bwzfXH iEmkta"
              x1="0%"
              y1="8px"
              x2="100%"
              y2="8px"
          />,
          }
        }
        key="0"
        x1="0%"
        x2="100%"
        y1="8px"
        y2="8px"
      >
        <line
          className="sc-bwzfXH iEmkta"
          x1="0%"
          x2="100%"
          y1="8px"
          y2="8px"
        />
      </styled.line>
      <styled.line
        innerRef={
          Object {
            "current": <line
              class="sc-bwzfXH iEmkta"
              x1="0%"
              y1="16px"
              x2="100%"
              y2="16px"
          />,
          }
        }
        key="1"
        x1="0%"
        x2="100%"
        y1="16px"
        y2="16px"
      >
        <line
          className="sc-bwzfXH iEmkta"
          x1="0%"
          x2="100%"
          y1="16px"
          y2="16px"
        />
      </styled.line>
      <styled.line
        innerRef={
          Object {
            "current": <line
              class="sc-bwzfXH iEmkta"
              x1="0%"
              y1="24px"
              x2="100%"
              y2="24px"
          />,
          }
        }
        key="2"
        x1="0%"
        x2="100%"
        y1="24px"
        y2="24px"
      >
        <line
          className="sc-bwzfXH iEmkta"
          x1="0%"
          x2="100%"
          y1="24px"
          y2="24px"
        />
      </styled.line>
    </svg>
  </styled.svg>
</MenuButton>
`;
